### PR TITLE
feat: introduce `dynamic()` and helper functions for dynamic colors

### LINKED ISSUE #2: Change colors dynamically afterwards
I really appreciate this plugin and use it with great pleasure.
However, there is one special feature I want to implement into my app.
I want to set the primary colour based on a colour picker that the user has in his settings menu.
Is there (or will there be) a usable solution that is compatible with the plugin?

`tailwind.config.js`
```js
import realtimeColors from 'tailwind-plugin-realtime-colors';

/** @type {import('tailwindcss').Config} */
export default {
	content: ['./src/**/*.{html,js,svelte,ts}'],
	theme: {
		extend: {
			fontFamily: {
				sans: ['Inter', 'sans-serif']
			}
		}
	},
	darkMode: 'class',
	plugins: [
		realtimeColors({
			colors: {
				text: '#f2ece8',
				background: '#040404',
				primary: '#2456df',
				secondary: '#000',
				accent: '#000'
			}
		})
	]
};
```

### PR DESCRIPTION
**Pull request by BlankParticle**
This introduces `dynamic()` as a way to skip generating a color and some helper functions to help in creation of dynamic colors

fixes #2

### CODE REVIEW

**Comment by finnmargraf** [comment]
Firstly, that type of implementation is exactly what I need 👍🏻 
But I have some open questions / bugs

1. CSS variable generation is creating duplicate rgba(). As you can see in the screenshot, the css vars from the generateDynamicPalette() function are inside a rgba() function, which (i think) collides with tailwind, since tw also puts every color variable into rgba(). Therefore, the color doesn't get displayed.
_Screenshot of css variables_
grafik
grafik

2. Non dynamic colors aren't generated at all. For some reason all my other, non-dynamic colors aren't generated at all. Maybe this is only in my app, I didn't test it quite well.

`tailwindcss.config.js`
```js
import realtimeColors from 'tailwind-plugin-realtime-colors';

/** @type {import('tailwindcss').Config} */
export default {
	content: ['./src/**/*.{html,js,svelte,ts}'],
	theme: {
		extend: {
			fontFamily: {
				sans: ['Inter', 'sans-serif']
			}
		}
	},
	darkMode: 'class',
	plugins: [
		realtimeColors({
			colors: {
				text: '#f2ece8',
				background: '#040404',
				primary: 'dynamic()',
				secondary: '#000',
				accent: '#000'
			}
		})
	]
};
```
`primaryColor.ts` (app specific)
```ts
import { browser } from '$app/environment';
import { generateDynamicPalette } from 'tailwind-plugin-realtime-colors';

export function setPrimaryColor(color: string) {
	if (!browser || !window) return;
	const palette = generateDynamicPalette({ primary: color });
	for (const [key, value] of Object.entries(palette)) {
		document.documentElement.style.setProperty(key, value);
	}
}
```

File: src/plugin.ts [hunk 1]
```diff
@@ -79,7 +89,10 @@ const getCSS = (config: RealtimeColorOptions) => {
   const modifiers = availableModifiers[config.shadeAlgorithm];
   const variables: Record<string, string> = {};
   const altVariables: Record<string, string> = {};
-  for (const [colorName, color] of Object.entries(colors)) {
+
+  for (const [colorName, color] of Object.entries(colors).filter(
+    ([_, color]) => color === "dynamic()",
```
Comment (finnmargraf) [inline src/plugin.ts]: Shouldn't it be `color !== "dynamic()"`?
Otherwise it would generate the colors for all **dynamic** colors?

**Comment by finnmargraf** [comment]
Also, one cool addition would be theming support. Probably an option in the generateDynamicPalette() to generate the inverted colors for dark mode would be the best way?

File: src/plugin.ts (hunk 1 above)
Comment (BlankParticle) [reply src/plugin.ts]: lol, good catch, maybe I removed the `!` before pushing

**Comment by BlankParticle** [comment]
> Also, one cool addition would be theming support. Probably an option in the generateDynamicPalette() to generate the inverted colors for dark mode would be the best way?

For that I have exported `invertColor`, as I cant ensure how the variables are gonna be used so I am handing that over to the user.

```ts
const primary = getPrimaryColor();
const lightColors = generateDynamicPalette({primary});
const darkColors = generateDynamicPalette(invertColor({primary}));

//Use them how you want
```

Edit: Maybe invert should also invert Objects with colors
Edit 2: I didn't add a option for invert in the function so that if you don't need the inverted colors it would be treeshaken.

**Comment by BlankParticle** [comment]
Also, I think `bg-primary-500/80` might work with dynamic on default config. Can you try that also.

Edit: can make sure works with `rgba`

**Comment by finnmargraf** [comment]
The palette output is still in the rgba() format, maybe I made a mistake or it is intended.
But nevertheless, the primary color doesn't get displayed in the app.

grafik

**Comment by BlankParticle** [comment]
> The palette output is still in the rgba() format, maybe I made a mistake or it is intended. But nevertheless, the primary color doesn't get displayed in the app.
> 
> grafik

that should not be the case, I just tried it myself. try deleting node_modules, pull latest changes and build it before using

Edit: It is not intended, it is a mistake form my side, not trying the plugin in a demo app. I will add tests soon.

**Comment by finnmargraf** [comment]
Okay it works fine when I build the app, I think it is the SvelteKit dependency optimization thing which causes some caching issues in the dev server build

**Comment by BlankParticle** [comment]
Happy to hear that, check if it solves everything you wanted. Then I will merge it and publish new version.
Otherwise you would have to wait for tomorrow afternoon (GMT+5:30)

**Comment by finnmargraf** [comment]
It should have everything I need. I just need to do a bit implementation on my side, but your code should be good to go

**Comment by finnmargraf** [comment]
Huge thanks @BlankParticle, I appreciate your work ❤️

**Pull request merged**